File: src/offer25.Merge_Sorted_Lists/merge_sorted_lists.cpp
/*
struct ListNode {
	int val;
	struct ListNode *next;
	ListNode(int x) :
			val(x), next(NULL) {
	}
};*/
class Solution1 {
public:
    ListNode* Merge(ListNode* pHead1, ListNode* pHead2)
    {
        if (pHead1 == nullptr)
            return pHead2;
        if (pHead2 == nullptr){
            return pHead1;
        }
        ListNode* mergeHead = nullptr;
        ListNode* cur = nullptr;
        while (pHead1 != nullptr && pHead2 != nullptr){
            if (pHead1->val <= pHead2->val){
                if (mergeHead == nullptr){
                    mergeHead = cur = pHead1;
                }
                else{
                    cur->next = pHead1;
                    cur = cur->next;
                }
                pHead1 = pHead1->next;
            }
            else{
                if (mergeHead == nullptr){
                    mergeHead = cur = pHead2;
                }
                else{
                    cur->next = pHead2;
                    cur = cur->next;
                }
                pHead2 = pHead2->next;
            }
        }
        if (pHead1 == nullptr){
            cur->next = pHead2;
        }
        else {
            cur->next = pHead1;
        }
        return mergeHead;
    }
};

/*
struct ListNode {
	int val;
	struct ListNode *next;
	ListNode(int x) :
			val(x), next(NULL) {
	}
};*/
class Solution2 {
public:
    ListNode* Merge(ListNode* pHead1, ListNode* pHead2)
    {
        if (pHead1 == nullptr)
            return pHead2;
        if (pHead2 == nullptr){
            return pHead1;
        }
        ListNode* pMergedHead = nullptr;
        if (pHead1->val < pHead2->val){
            pMergedHead = pHead1;
            pMergedHead->next = Merge(pHead1->next, pHead2);
        }
        else{
            pMergedHead = pHead2;
            pMergedHead->next = Merge(pHead1, pHead2->next);
        }
        return pMergedHead;
    }
};

File: src/offer25.Merge_Sorted_Lists/merge_sorted_lists.java
//1
/*
public class ListNode {
    int val;
    ListNode next = null;

    ListNode(int val) {
        this.val = val;
    }
}*/
public class Solution {
    public ListNode Merge(ListNode list1,ListNode list2) {
        ListNode newlist = new ListNode(-1);
        ListNode head = newlist;
        while(list1 != null && list2 != null){
            if(list1.val < list2.val){
                newlist.next = new ListNode(list1.val);
                list1 = list1.next;
                newlist = newlist.next;
            }else{
                newlist.next = new ListNode(list2.val);
                list2 = list2.next;
                newlist = newlist.next;
            }
        }
        while(list1 != null){
            newlist.next = new ListNode(list1.val);
            list1 = list1.next;
            newlist = newlist.next;
        }
        
        while(list2 != null){
            newlist.next = new ListNode(list2.val);
            list2 = list2.next;
            newlist = newlist.next;
        }
        return head.next;
    }
}
//2
/*
public class ListNode {
    int val;
    ListNode next = null;

    ListNode(int val) {
        this.val = val;
    }
}*/
public class Solution {
    public ListNode Merge(ListNode list1,ListNode list2) {
        if(list1 == null){
            return list2;
        }
        if(list2 == null){
            return list1;
        }
        ListNode mergedNode = null;
        if(list1.val < list2.val){
            mergedNode = list1;
            mergedNode.next = Merge(list1.next, list2);
        }else{
            mergedNode = list2;
            mergedNode.next = Merge(list1, list2.next);
        }
        return mergedNode;
    }
}